### PR TITLE
Implement Wasm GC in Wasmtime

### PR DESCRIPTION
**Pull request by fitzgen**
Implement support for [the WebAssembly garbage collection proposal][wasm-gc-proposal] in Wasmtime. Allow plugging different collectors into Wasmtime to support the requirements of different Wasmtime embeddings running different Wasm workloads.

[wasm-gc-proposal]: https://github.com/WebAssembly/gc

### [Rendered](https://github.com/fitzgen/rfcs-2/blob/wasm-gc/accepted/wasm-gc.md)

### CODE REVIEW

**Comment by c-cube** [comment]
Hi, just a drive-by suggestion: has [Immix](https://www.cs.cornell.edu/courses/cs6120/2019fa/blog/immix/) been considered as an alternative to Boehm?  It's a precise GC afaik, whereas Boehm is conservative.

**Comment by programmerjake** [comment]
imho wasm gc will eventually need to support a multi-threaded wasm program with gc, so do try to account for this and don't rely on wasmtime api types being `!Send` unless you can think of a reasonable way to access them from a different thread (e.g. separate `Send + Sync` and `!Send` types, where the latter is kinda like a mutex guard)

**Comment by waywardmonkeys** [comment]
@fitzgen Have you seen or looked at https://github.com/Ravenbrook/mps ?

@rptb1 may have more to add, but MPS is a great GC and has been used in production scenarios over the years. There's not a Rust binding at the moment, but that seems like it could be resolved.

File: accepted/wasm-gc.md [hunk 1]
```diff
@@ -0,0 +1,1319 @@
...
+* When attempting to access information about an array, struct, or function type
+  whose sole defining module has been dropped, should we
+
+  * panic, or
+
+  * return `None` (meaning that all these type accessors return `Option`s)
+
+  ?
+
+* Should `AnyRef` and `EqRef` be collapsed together in the public API? Right now
```
Comment (tlively) [inline accepted/wasm-gc.md:1304]: If the embedder so chooses, it is possible for internalized external references to inhabit `anyref` without inhabiting `eqref`. For example, in the JS embedding, most internalized references to JS objects are not `eqref`. Something similar might make sense for wasmtime.

More generally, how do you plan to implement `extern.externalize` and `extern.internalize`? I didn’t see that discussed in the RFC.

**Comment by fitzgen** [comment]
@programmerjake

> imho wasm gc will eventually need to support a multi-threaded wasm program with gc, so do try to account for this and don't rely on wasmtime api types being `!Send` unless you can think of a reasonable way to access them from a different thread (e.g. separate `Send + Sync` and `!Send` types, where the latter is kinda like a mutex guard or maybe a `Rc<Arc<T>>`)

Minor correction: the relevant `wasmtime` API types are `!Sync + Send`, not `!Send`.

Regarding the larger idea of combining Wasm threads and Wasm GC: every sketch I've seen about extending the Wasm threading model to new bits of state requires marking that state as `shared` the same way that memories are either `shared` or not. So GC types would need to be marked `shared` to allow multi-threaded access to their instances. Assuming that is the case, then when we are compiling Wasm code, we'll be able to only emit atomic and synchronized operations when necessary and omit them when we the type information tells us it is statically unnecessary. In the public `wasmtime` API we may want to have separate Rust types for reflecting `shared` vs not Wasm GC types, the way we have `wasmtime::SharedMemory` and `wasmtime::Memory`, or allow dynamic downcasting to unsynchronized types based on checking whether the underlying GC object's type is `shared` or not. But all this is a bit speculative and far future as there hasn't been any proposal about how to combine these two different features yet, as far as I'm aware.

File: accepted/wasm-gc.md (hunk 1 above)
Comment (fitzgen) [reply accepted/wasm-gc.md:1304]: Ah, this is a great point, thanks @tlively. I overlooked those instructions and their implications. I think they do force our hand in two ways:

1. We need to manage `externref`s and `anyref`s with the same collector, since it will be trivial to create cycles between the two, since `externref`s aren't solely created by the host anymore.
2. We do need `wasmtime::AnyRef` that is separate from `wasmtime::EqRef` so that we can represent host-created `externref`s that have been internalized and which do not support equality comparison.

Thanks for helping clear these things up!

As for how to implement the instructions, everything will probably be easier if we adopt a uniform representation between `anyref` and `externref`, but I haven't had a chance to give the problem too much thought yet :)

File: accepted/wasm-gc.md [hunk 2]
```diff
@@ -0,0 +1,1319 @@
...
+    /// object headers it requires.
+    fn rec_group_layouts(&mut self, sub_types: &[wasm_types::SubType]) -> Vec<GcObjectLayout>;
+
+    /// Return the CLIF type for GC references with this collector. Must be
+    /// either `r32` or `r64`.
+    fn reference_type(&mut self, isa: &dyn TargetIsa) -> ir::types::Type;
+
+    /// The function builder type for this compiler (that is,
+    /// `cranelift_frontend::FunctionBuilder` or the Winch equivalent).
+    type FunctionBuilder;
```
Comment (alexcrichton) [inline accepted/wasm-gc.md:629]: FWIW this'll be funky because this make `GcCompiler` not object safe

File: accepted/wasm-gc.md [hunk 3]
```diff
@@ -0,0 +1,1319 @@
...
+call and add this scheme later if performance becomes a concern.
+
+## GC Implementations
+
+We will have three GC implementations:
+
+1. The null GC, which allocates until it runs out of heap space, at which
+   point it triggers a trap in the guest.
+
+2. A general-purpose collector, built on top of an off-the-shelf GC library.
```
Comment (alexcrichton) [inline accepted/wasm-gc.md:1032]: Out of curiosity, is the "GC" part of the off-the-shelf library necessary? It seems like we're doing all the hard work of rooting and maintaining layouts, so building an implementation on top of `malloc` wouldn't really be all that much worse than any other GC library at that point (like boehm or something like that I'd presume).

This definitely makes sense to have, but it seems to be like it could be basically something backed by `SystemAlloc` in Rust.

**Review by alexcrichton — COMMENTED** [review]
This all sounds great to me!

I read the term "GC pause" in relation to the pooling instance allocator and something it reminded me of is the interaction with epochs/fuel. Embedders limiting the execution time of a wasm instance will often use fuel for determinism, epochs for a timeout, or epochs plus async for "allow me to cancel at any epoch time slice". If GC is expected to take a negligible amount of time (e.g. short-lived instances) this may not be much of a problem but as the cost of a GC grows this may become noticeable.

I'm not really sure what to do with this myself, but we may end up needing to integrate epochs and fuel into the stop-the-world part of the GC. That would necesitate some level of "save the work for later and then resume" style GC's though which doesn't seem like it would be trivial. Alternatively though, especially for a time-sliced HTTP server, pausing for a "big GC" also seems like it could end quite badly too. 

Basically another way to put this is that I'm not sure if there's sufficient knobs for an embedder to protect themselves against a pathological scenario where GCs are frequent and very long. Do you have thoughts though on how this could be addressed?

File: accepted/wasm-gc.md (hunk 1 above)
Comment (fitzgen) [reply accepted/wasm-gc.md]: @tlively I added some notes in the most recent commit about these instructions, if you're interested in my current thinking on them.

Comment (tlively) [reply accepted/wasm-gc.md]: Thanks! Looks good to me.

File: accepted/wasm-gc.md [hunk 4]
```diff
@@ -0,0 +1,1347 @@
...
+```
+
+Note that the `StructRef::{get,set}` and `ArrayRef::{get,set}` methods require
+passing in a mutable `wasmtime::Store` context. This is because the GC objects
+are only valid while the store is live and the whole GC heap may be deallocated,
+reclaimed, or otherwise recycled for new instances once the store and its Wasm
+instances are dropped. Therefore, these methods cannot simply dereference
+pointers to their underlying objects, they need to assert that the given context
+is the same store that their object came from, which means the referenced object
+is still live, because `wasmtime::Ref`s always root there referent.
```
Comment (vados-cosmonic) [inline accepted/wasm-gc.md:446]: s/there/their/ ?

Comment (fitzgen) [reply accepted/wasm-gc.md]: Fixed, thanks!

File: accepted/wasm-gc.md [hunk 5]
```diff
@@ -0,0 +1,1319 @@
...
+    // ...
+}
+
+impl From<EqRef> for AnyRef { /* ... */ }
+
+impl TryFrom<AnyRef> for EqRef { /* ... */ }
+
+impl AnyRef {
+    pub fn ref_type(&self) -> RefType { /* ... */ }
+    pub fn is_instance_of(&self, ty: RefType) -> bool { /* ... */ }
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply accepted/wasm-gc.md:310]: Yeah I was imagining that these would hold a `RefType` internally but general `is_instance_of` calls will need a store/engine to look up indexed type info (such as `ArraryType(...)` as opposed to like `HeapType::EqRef`).

We could shrink these structs a little by pushing the type into the store as well, but that would make even the simplest accessors like `AnyRef::ref_type` require a store, which is maybe a slight inconvenience but also not that different from most of the `wasmtime` public API. Could go either way.

File: accepted/wasm-gc.md [hunk 6]
```diff
@@ -0,0 +1,1319 @@
...
+mutability) when we have mutable access to a store (e.g. when allocating new
+entries or when collecting garbage). Finally, because
+`wasmtime::{Struct,Array}Ref`s don't internally hold a mutable store, when they
+decrement their associated root set entry's ref count and it reaches zero, they
+cannot remove the zero-ref-count entry from the set. That is delayed until
+either the next garbage collection or when the whole store is dropped.
+
+Altogether, this design for exposing reference values in the public API allows
+us to keep performance-degrading synchronization to a minimum while still
+providing a fully safe and relatively ergonomic API.
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply accepted/wasm-gc.md]: I've pushed a commit to make the RFC more upfront about these costs in the host API. We could do something like you describe, but I think we won't have a ton of pressure on this level of performance in the host API for a while.

File: accepted/wasm-gc.md (hunk 2 above)
Comment (fitzgen) [reply accepted/wasm-gc.md:672]: I think not being object safe will actually be okay since we monomorphize based on func env anyways. Will update the part of the RFC that assumes object safety.

File: accepted/wasm-gc.md (hunk 3 above)
Comment (fitzgen) [reply accepted/wasm-gc.md:1076]: The hard part is not rooting and object layouts, but implementing a nursery, compaction, background sweeping, parallel tracing, incremental GC, and everything else that a general-purpose GC is going to eventually want.

File: accepted/wasm-gc.md [hunk 7]
```diff
@@ -0,0 +1,1319 @@
...
+
+impl PartialEq for ArrayRef { /* ... */ }
+impl Eq for ArrayRef { /* ... */ }
+
+impl ArrayRef {
+    pub fn null(ty: ArrayType) -> ArrayRef { /* ... */ }
+
+    pub fn new(
+        store: impl AsContextMut,
+        ty: StructType,
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply accepted/wasm-gc.md]: I think this one might have gotten hidden in the other changes

File: accepted/wasm-gc.md [hunk 8]
```diff
@@ -0,0 +1,1362 @@
...
+impl PartialEq for ArrayRef { /* ... */ }
+impl Eq for ArrayRef { /* ... */ }
+
+impl ArrayRef {
+    pub fn null(ty: ArrayType) -> ArrayRef { /* ... */ }
+
+    pub fn new(
+        store: impl AsContextMut,
+        ty: ArrayType,
+        elements: impl Iterator<Item = Val>,
```
Comment (jakobkummerow) [inline accepted/wasm-gc.md:424]: For implementing `array.new` and `array.new_default`, it may be useful to have a way to construct `ArrayRef` from a single element value, which will be used for all fields. Or maybe constructing an iterator that returns the same value again and again is so easy that you can just rely on that.

File: accepted/wasm-gc.md [hunk 9]
```diff
@@ -0,0 +1,1362 @@
...
+    Struct(GcStructLayout),
+    Array(GcArrayLayout),
+}
+
+/// The layout for a GC-managed struct type.
+pub struct GcStructLayout {
+    /// The size of this struct.
+    pub size: u32,
+    /// The alignment of this struct.
+    pub align: u32,
```
Comment (jakobkummerow) [inline accepted/wasm-gc.md:636]: In V8, object alignment is a global constant (and equals the pointer size), rather than being stored per object layout. That makes some things simpler, but sometimes we do wish we had the infrastructure to have per-object alignment requirements, so it'll be interesting to see whether you find that having this flexibility is worthwhile.

File: accepted/wasm-gc.md [hunk 10]
```diff
@@ -0,0 +1,1362 @@
...
+is always rooted, it keeps everything it can reach alive until dropped, and
+because there is a cycle it won't ever be dropped.
+
+This RFC proposes that we do not initially support these features and not having
+them is acceptable for the time being. If and when there is sufficient demand
+and motivation, we can add a `GcTrace` trait and unrooted reference type
+later. There will be more design work to be done, but it won't fundamentally
+change any of the rest of this RFC's design.
+
+## Implementing `extern.externalize` and `extern.internalize`
```
[the remark this answers is not in the document]
Comment (tlively) [reply accepted/wasm-gc.md:533]: > because internalized externrefs aren't eqrefs.

IIUC, this is largely up to the embedding. For example, in the JS embedding, internalized numbers may be i31refs, which are eqrefs. I believe different embeddings are allowed to do different things here, such as always internalizing to eqref or even to specific struct and array types.

This flexibility could be a portability hazard, so I've opened https://github.com/WebAssembly/gc/issues/369 to discuss it.

File: accepted/wasm-gc.md (hunk 3 above)
Comment (alexcrichton) [reply accepted/wasm-gc.md:1076]: (still have some thoughts so un-resolving this temporarily)

For a first-pass implementation do we need that though? Of the benefits you listed:

* Nursery / compaction - these seem like things that `malloc` is a "proxy" for, although I realize that `malloc` is not perfect for these of course. 
* Background sweeping - we're not going to want this any time in the near future I think? That'd be reliant on something like the `shared` attribute in core wasm?
* Parallel tracing - does a GC library do this for us? Wouldn't we do all the tracing ourselves since we're the ones understanding the heap layout of all GC objects? I haven't worked with any off-the-shelf GC libraries so I'm likely missing something here.

I can understand how perhaps, one day, we want a super-fancy all-the-bells-and-whistles GC. That to me feels on the order of numbers-of-years out though so it seems like "just call `malloc`" with a simple mark/sweep could be sufficient for quite some time.

File: accepted/wasm-gc.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexcrichton) [reply accepted/wasm-gc.md]: Sounds reasonable to me 👍

**Comment by fitzgen** [comment]
@alexcrichton 

> I'm not really sure what to do with this myself, but we may end up needing to integrate epochs and fuel into the stop-the-world part of the GC. That would necesitate some level of "save the work for later and then resume" style GC's though which doesn't seem like it would be trivial. Alternatively though, especially for a time-sliced HTTP server, pausing for a "big GC" also seems like it could end quite badly too.
> 
> Basically another way to put this is that I'm not sure if there's sufficient knobs for an embedder to protect themselves against a pathological scenario where GCs are frequent and very long. Do you have thoughts though on how this could be addressed?

I don't think this actually requires that the GC is incremental the way that the GC literature defines it, because these scenarios don't require resuming the mutator when we async yield.

But that is a good point that we want the equivalent of

```rust
trait GcRuntime {
    // ...

    unsafe async fn async_gc(
        &self,
        heap: GcHeap,
        roots: RootSet,
        yield_every: std::time::Duration,
    );
}
```

but of course it would be a bit uglier than that because we can't have `async` methods in traits yet.

I think this should be sufficient to address your concerns, yeah?

File: accepted/wasm-gc.md (hunk 3 above)
Comment (fitzgen) [reply accepted/wasm-gc.md:1088]: > Nursery / compaction - these seem like things that malloc is a "proxy" for, although I realize that malloc is not perfect for these of course.

Do you mean proxy in that the `malloc` implementation should handle these things for us? I don't think it can or will. I'm not totally following you here.

> Background sweeping - we're not going to want this any time in the near future I think? That'd be reliant on something like the shared attribute in core wasm?

Because wasm-gc doesn't have finalizers, it isn't semantically visible to the wasm program where/when sweeping happens. Therefore, it doesn't rely on anything like a `shared` attribute, and is purely a runtime optimization. And it is a very nice optimization, since sweeping is O(heap size) not O(live objects), at least in simple mark-sweep.

> Parallel tracing - does a GC library do this for us? Wouldn't we do all the tracing ourselves since we're the ones understanding the heap layout of all GC objects? I haven't worked with any off-the-shelf GC libraries so I'm likely missing something here.

It depends on the details of the library, I suppose. I haven't used any of these libraries in anger myself either.

But before digging in too much further: we are *not* the ones who understand object layout. It is the collector that knows object layout, and us who have to indirectly access fields. And object layout *has* to be the collector's responsibility since different collectors have different requirements for things like header words or reference counts or any other kind of metadata it requires.

> I can understand how perhaps, one day, we want a super-fancy all-the-bells-and-whistles GC. That to me feels on the order of numbers-of-years out though so it seems like "just call malloc" with a simple mark/sweep could be sufficient for quite some time.

Yeah, perhaps if we don't hear from any of the non-FaaS Wasmtime embedders (who are all going to want to use the semi-space copying collector) we can ignore the general-purpose GC usecase for now and only have the null collector and the semi-space copying collector.

But if we still want to have a general-purpose GC, then I'm not sure what you are proposing. Just doing `malloc`/`free` and simple mark-sweep ourselves is going to be a very suboptimal collector, probably worse than the semi-space copying collector for long-running Wasm programs. So I'm not sure it would actually gain us anything.

File: accepted/wasm-gc.md [hunk 11]
```diff
@@ -0,0 +1,1362 @@
...
+`wasmtime::ValType` and into `wasmtime::RefType`:
+
+```rust
+pub enum ValType {
+    I32,
+    I64,
+    F32,
+    F64,
+    V128,
+    Ref(RefType),
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply accepted/wasm-gc.md:98]: Thanks for the notes @jakobkummerow! Really appreciate getting feedback from another implementor.

Can you expand a bit on why moving the split "up" into `ValType` was useful?

File: accepted/wasm-gc.md [hunk 12]
```diff
@@ -0,0 +1,1362 @@
...
+
+```rust
+pub enum Val {
+    // <existing variants>
+
+    Ref(Ref),
+}
+
+pub enum Ref {
+    Internal(AnyRef),
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply accepted/wasm-gc.md:263]: Right now all `{Any,Eq,Struct,Array}Ref`s are nullable on the Rust side.

That is, as written we don't represent null with `None` and non-null with `Some(FooRef)` and non-nullable foo refs with `FooRef`.

This is a good point to consider though. Should we switch to that representation? It is actually what we do with `funcref` and `externref` right now in the public API, but it does mean that you can't get precise type information for a null reference. I guess that is maybe not a super useful property to have though.

@alexcrichton do you happen to have thoughts on this?

File: accepted/wasm-gc.md [hunk 13]
```diff
@@ -0,0 +1,1362 @@
...
+    pub fn into_array_ref(self) -> Option<ArrayRef> { /* ... */ }
+}
+```
+
+Note that the `into_*_ref` methods avoid reference counting operations. More one
+this below.
+
+`I31Ref` allows getting its inner 31-bit value as either a signed `i32` or
+unsigned `u32` and creation from those types as well (with a check that the
+value fits in 31 bits):
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply accepted/wasm-gc.md:357]: I'll add masking constructor variants. I can imagine there are scenarios where you want to check vs don't want to, and checking by default feels more Rust-y to me, so I don't want to get rid of those variants.

File: accepted/wasm-gc.md (hunk 8 above)
Comment (fitzgen) [reply accepted/wasm-gc.md:439]: Great idea, thanks!

File: accepted/wasm-gc.md [hunk 14]
```diff
@@ -0,0 +1,1362 @@
...
+entries or when collecting garbage). Finally, because
+`wasmtime::{Struct,Array}Ref`s don't internally hold a mutable store, when they
+decrement their associated root set entry's ref count and it reaches zero, they
+cannot remove the zero-ref-count entry from the set. That is delayed until
+either the next garbage collection or when the whole store is dropped.
+
+Altogether, this design for exposing reference values in the public API allows
+us to keep performance-degrading synchronization to a minimum while still
+providing a fully safe and relatively ergonomic API and simultaneously keeping
+Wasm manipulation of GC objects as fast as possible.
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply accepted/wasm-gc.md:513]: Great, thanks for the validation!

File: accepted/wasm-gc.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (fitzgen) [reply accepted/wasm-gc.md:548]: Yes, we allow creating `externref`s from arbitrary host things, although if you create an `externref` of a `String` for example, we don't pass a `*mut String` into Wasm. We always allocate an `ExternRef` wrapper object today, and I don't see that changing in the future.

File: accepted/wasm-gc.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jakobkummerow) [reply accepted/wasm-gc.md:98]: It allows us to have an encoding that's both simple and memory-efficient: our `ValueType` class is internally a 32-bit field that encodes a "kind" (i8, i16, i32, i64, f32, f64, ref, refnull, rtt, void, bottom), and a type index for those kinds that have one. We even had enough bits for the RTT depth back when we needed to store it.
There is of course ample design room to make different choices about the details, I'm just saying that the simple two-level design has served us well so far. And to be fair, I don't know how the requirements will change if/when we get future type system features such as generics or union types or whatnot.

File: accepted/wasm-gc.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jakobkummerow) [reply accepted/wasm-gc.md:263]: I think the useful ability is to avoid null checks when you statically know that a value going into an operation is non-nullable anyway. I think all instructions accept nullable values, so you can probably get away with not expressing nullability in your value representation, but why leave performance on the table?

**Comment by alexcrichton** [comment]
I'm not sure myself really, I don't feel like I've got a good enough handle on the problems that GC might bring pause-wise. There's a balancing act between what the GC algorithms can support and what embedders want to happen. For example can all GC algorithms easily implement what you listed above where, in the middle of a GC, it realizes "time's up" and records info to resume later? I'm not sure additionally how it would work to thread through the deadline into embedders. For example they could configure "take no longer than 2ms" but if WebAssembly already nearly consumed its entire time slice with executing code and then the GC takes an additional 2ms is that bad? 

I'm not sure that we're really positioned to answer these questions at this time. I feel like we'll basically have to wait and see and figure it out as we go along and have more constraints available to us to design a solution within.

File: accepted/wasm-gc.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (alexcrichton) [reply accepted/wasm-gc.md:263]: > but it does mean that you can't get precise type information for a null reference

I'm not sure what you mean by this @fitzgen? Otherwise I'd naively say that `Option<T>` would be the better representation than internally storing the fact that it's null. I may not understand what else is in play here, but I would assume that the precise type information for the reference comes from the `T`?

File: accepted/wasm-gc.md (hunk 3 above)
Comment (alexcrichton) [reply accepted/wasm-gc.md:1103]: Hm before I answer some other things one thing you said confuses me:

> And object layout has to be the collector's responsibility since different collectors have different requirements for things like header words or reference counts or any other kind of metadata it requires.

I don't understand this because doesn't compiled code need to access the object's layout? Doesn't this mean that, no matter what, the in-memory layout of an object is intimately know by the runtime to something other than a collector as a black box? For example the `GcCompiler` trait would have a very close relationship to the collector, so if we picked an off-the-shelf library doing fancy things here we'd reverse-engineer it and codify it in a corresponding `GcCompiler` implementation. 

I think overall what I'm feeling is that we should start with something backed by mark-and-sweep plus `malloc` knowing that it's not going to be that performant. We can have a specialization for the pooling allocator and FaaS-style runtimes of course, but I'm hesitant to bring in a large GC library dependency. GC is notoriously hard to get right and I'm not sure that any of us are so familiar with any particular GC library that we'd be willing to stake our security on the correctness of it and our understanding of its internals.

Comment (fitzgen) [reply accepted/wasm-gc.md:1103]: > I don't understand this because doesn't compiled code need to access the object's layout? Doesn't this mean that, no matter what, the in-memory layout of an object is intimately know by the runtime to something other than a collector as a black box? For example the GcCompiler trait would have a very close relationship to the collector, so if we picked an off-the-shelf library doing fancy things here we'd reverse-engineer it and codify it in a corresponding GcCompiler implementation.

Yes, object layout would be completely encapsulated within `GcCompiler` implementations. The `GcCompiler` implementation wouldn't really "reverse engineer" the object layout, but would be considered "part of" the collector and would have insider knowledge/collaboration that the rest of the compiler/runtime wouldn't have (or would only get indirectly via `GcObjectLayout`) and which would potentially differ across different collector implementations.

> I think overall what I'm feeling is that we should start with something backed by mark-and-sweep plus malloc knowing that it's not going to be that performant. We can have a specialization for the pooling allocator and FaaS-style runtimes of course, but I'm hesitant to bring in a large GC library dependency. GC is notoriously hard to get right and I'm not sure that any of us are so familiar with any particular GC library that we'd be willing to stake our security on the correctness of it and our understanding of its internals.

I agree with you about the difficulty of building a correct and performant GC.

What I'm saying though is that if we don't want to pull in an external library for the general-purpose GC use case, then we shouldn't attempt to build our own. Instead we should just use the semi-space copying collector that we need for pooling allocator integration for the general-purpose GC use case. It won't be as good as what someone could build with MMTK, but it will be better than naive mark-and-sweep backed by malloc/free. And this approach allows us to build just two GCs (null, semi-space copying) instead of three (null, semi-space copying, mark-and-sweep).

**Comment by fitzgen** [comment]
> I'm not sure myself really, I don't feel like I've got a good enough handle on the problems that GC might bring pause-wise. There's a balancing act between what the GC algorithms can support and what embedders want to happen. For example can all GC algorithms easily implement what you listed above where, in the middle of a GC, it realizes "time's up" and records info to resume later? I'm not sure additionally how it would work to thread through the deadline into embedders. For example they could configure "take no longer than 2ms" but if WebAssembly already nearly consumed its entire time slice with executing code and then the GC takes an additional 2ms is that bad?
> 
> I'm not sure that we're really positioned to answer these questions at this time. I feel like we'll basically have to wait and see and figure it out as we go along and have more constraints available to us to design a solution within.

I don't think we need to require that all collectors to implement `async_gc`, just the ones we intend to use with async, ie the collector that integrates with the pooling allocator. But in general, production GCs are very used to trying to minimize stop-the-world pauses and doing work incrementally. Our problem is easier than usual because we aren't even resuming the mutator in between slices. But you're right that this might not align with existing off-the-shelf libraries and their APIs. I think that is fine: the same way that some collectors just won't integrate with the pooling allocator, some won't support async collection. And if we move to having just the null GC and the semi-space copying collector (as discussed up thread) then we can make them both support async pretty easily.

As far as the API for setting quanta and all that: I'm all ears. Rather than a single `Future` that completes when GC is finished, we could have many futures, each completing when their time slice is completed and then returning either "I'm done GCing, here is your instance back" or "I have more GC work to do, give me a quantum and I'll return a new future for this next time slice". This would let us have different lengths of GC slices and notably make that first one shorter because of the amount of time that Wasm has already executed for the whole store's current time slice. Or we could explicitly pass in some kind of "check whether you should yield" handle to the collector that checks the epoch in `VMRuntimeLimits` directly. I kind of like this latter approach...

File: accepted/wasm-gc.md (hunk 3 above)
Comment (fitzgen) [reply accepted/wasm-gc.md]: I've updated the RFC to remove MMTK and the general-purpose collector. Now there is just the null GC and the semi-space copying collector.

Comment (alexcrichton) [reply accepted/wasm-gc.md]: Ok yeah that all sounds reasonable to me 👍

**Comment by fitzgen** [comment]
# Motion to Finalize

**Disposition: Merge**

It seems like we have mostly come to consensus here and the bits that we haven't hammered out 100% are things we can do during the implementation phase. As such, I am moving that we merge this RFC.

------------------

Note: I don't want to trawl through commit logs and try and figure out which committer is part of which BA organization, so I'm not making a stakeholder check list. Just approve the PR or comment to sign off, or alternatively leave a comment detailing your objections if you have them. I'll manually keep track of whether we have multi-org sign off and all that as the comments come in.

As always, details on the RFC process can be found here: https://github.com/bytecodealliance/rfcs/blob/main/accepted/rfc-process.md#making-a-decision-merge-or-close

**Review by alexcrichton — APPROVED** [review]

**Review by cfallin — APPROVED** [review]

**Review by pchickey — APPROVED** [review]
Great work, Nick!

**Review by tschneidereit — APPROVED** [review]
This is really great! <3

**Comment by fitzgen** [comment]
As there have been signoffs from at least two people representing different stakeholder organizations, this RFC is entering its 10 day

# Final Comment Period

and the last day to raise objections before this can merge is 2023-06-03.

**Comment by fitzgen** [comment]
Seeing as no objections have been raised during the final comment period, I'm going to go ahead and merge this. Thanks everyone!
